File: messaging/kafka/java/eos/src/main/java/tech/khorost/kafka/eos/Txn.java
package tech.khorost.kafka.eos;

import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.clients.consumer.ConsumerRecords;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.apache.kafka.common.serialization.StringSerializer;

import java.time.Duration;
import java.util.List;
import java.util.Properties;

/**
 * (1)/(2) ядро транзакций: транзакционный producer, батч A коммитится
 * штатно, батч B на первой попытке абортится (симулированный сбой в
 * середине), пересылается повторно и коммитится. Тот же сценарий, что
 * ../../../java-deep-dive/messaging/.../EosDemo.java (не переиспользуется
 * напрямую — этот стенд разбит на CLI-фазы, как replication/storage), и то
 * же поведение, что ../../go/eos/txn.go — franz-go/kafka-clients сходятся
 * побайтово в физическом/логическом счёте.
 */
public final class Txn {
    private Txn() {
    }

    private static KafkaProducer<String, String> newTxnProducer(String txnId) {
        Properties props = new Properties();
        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, Kafka.BOOTSTRAP);
        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
        props.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, txnId);
        // С transactional.id заданным kafka-clients берёт idempotence=true по
        // дефолту; а явная попытка enable.idempotence=false + transactional.id
        // ПАДАЕТ с ConfigException("Cannot set a transactional.id without also
        // enabling idempotence.") — симметрично franz-go, который так же
        // запрещает DisableIdempotentWrite() при заданном TransactionalID.
        props.put(ProducerConfig.ACKS_CONFIG, "all");
        KafkaProducer<String, String> producer = new KafkaProducer<>(props);
        // ⚠️ Явный initTransactions() — ОБЯЗАТЕЛЕН в Java API (в отличие от
        // franz-go, где BeginTransaction() сам восстанавливает producer id
        // внутри себя). initTransactions() тоже фенсит/абортит зависшую
        // транзакцию ПРЕДЫДУЩЕГО процесса с тем же transactional.id (KIP-98) —
        // именно на этот вызов и рассчитывает попытка B в cpp-сценарии.
        producer.initTransactions();
        return producer;
    }

    public static final class BatchResult {
        int sentPhysically;
        int committedLogical;
    }

    private static int sendBatch(KafkaProducer<String, String> producer, String topic, String prefix, int batchSize, boolean failMidway) {
        producer.beginTransaction();
        int produced = 0;
        RuntimeException midwayErr = null;
        for (int i = 0; i < batchSize; i++) {
            // ⚠️ Синхронный .get() (ждём ack брокера) НАРОЧНО, а не обычный
            // асинхронный send() — если оставить асинхронно, найденный живьём
            // факт: abortTransaction() аварийно отменяет ЕЩЁ НЕ
            // ПОДТВЕРЖДЁННЫЕ брокером записи клиентски (они физически на
            // диск вообще не попадают, Javadoc KafkaProducer.abortTransaction:
            // "Any unflushed produce messages will be aborted..."), и
            // read_uncommitted-счёт физического мусора от абортнутого батча
            // тогда оказывается 0, а не 3 — воспроизведено живьём при первой
            // версии этого стенда. Franz-go в
            // ../../go/eos/txn.go использует ProduceSync (тоже синхронно) —
            // чтобы сравнение "физически записано" было яблоки-к-яблокам
            // между клиентами, а не артефактом разной буферизации, здесь
            // синхронный .get() тоже.
            try {
                producer.send(new ProducerRecord<>(topic, prefix + "-key-" + i, prefix + "-" + i)).get();
            } catch (Exception e) {
                throw new RuntimeException("send " + prefix + " i=" + i, e);
            }
            produced++;
            if (failMidway && i == 2) {
                midwayErr = new RuntimeException("симулированный сбой обработки в середине батча " + prefix);
                break;
            }
        }
        if (midwayErr != null) {
            producer.abortTransaction();
            System.out.printf("  [txn] батч %s: сбой (%s) -> abortTransaction() — %d записей физически ушли, логически ничего не подтверждено%n",
                    prefix, midwayErr.getMessage(), produced);
            return -produced; // отрицательное значение = абортнутый батч
        }
        producer.commitTransaction();
        System.out.printf("  [txn] батч %s: commitTransaction() — %d сообщений подтверждено%n", prefix, produced);
        return produced;
    }

    public static BatchResult runTxnBatches(String topic, int batchSize) {
        BatchResult result = new BatchResult();
        try (KafkaProducer<String, String> producer = newTxnProducer("cookbook-eos-txn-producer-java")) {
            int a = sendBatch(producer, topic, "batchA", batchSize, false);
            result.sentPhysically += Math.abs(a);
            if (a > 0) result.committedLogical += a;

            int b1 = sendBatch(producer, topic, "batchB", batchSize, true);
            result.sentPhysically += Math.abs(b1);
            if (b1 > 0) result.committedLogical += b1;

            int b2 = sendBatch(producer, topic, "batchB", batchSize, false);
            result.sentPhysically += Math.abs(b2);
            if (b2 > 0) result.committedLogical += b2;
        }
        System.out.printf("[txn] физически отправлено записей: %d, логически подтверждено: %d%n",
                result.sentPhysically, result.committedLogical);
        return result;
    }

    // package-private (не private) — переиспользуется из Cpp.cppVerify для
    // счёта output read_committed/read_uncommitted тем же кодом.
    static int consumeIsolation(String topic, String isolationLevel, long idleTimeoutMs) {
        Properties props = new Properties();
        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, Kafka.BOOTSTRAP);
        props.put(ConsumerConfig.GROUP_ID_CONFIG, "eos-txn-verify-" + isolationLevel + "-" + System.nanoTime());
        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
        props.put(ConsumerConfig.ISOLATION_LEVEL_CONFIG, isolationLevel);
        props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false");

        int count = 0;
        try (KafkaConsumer<String, String> consumer = new KafkaConsumer<>(props)) {
            consumer.subscribe(List.of(topic));
            long lastProgress = System.currentTimeMillis();
            while (true) {
                ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(500));
                count += records.count();
                if (!records.isEmpty()) {
                    lastProgress = System.currentTimeMillis();
                } else if (System.currentTimeMillis() - lastProgress > idleTimeoutMs) {
                    break;
                }
            }
        }
        return count;
    }

    public static void runTxnVerify(String topic, int expectCommitted, int expectPhysical) {
        int committedCount = consumeIsolation(topic, "read_committed", 5000);
        int uncommittedCount = consumeIsolation(topic, "read_uncommitted", 5000);
        System.out.printf("[txn-verify] read_committed=%d read_uncommitted=%d (ожидалось: committed=%d physical=%d)%n",
                committedCount, uncommittedCount, expectCommitted, expectPhysical);

        if (committedCount != expectCommitted) {
            throw new IllegalStateException("[txn-verify] РАСХОЖДЕНИЕ: read_committed=" + committedCount + ", ожидалось " + expectCommitted);
        }
        if (uncommittedCount != expectPhysical) {
            throw new IllegalStateException("[txn-verify] РАСХОЖДЕНИЕ: read_uncommitted=" + uncommittedCount + ", ожидалось " + expectPhysical);
        }
        if (!(uncommittedCount > committedCount)) {
            throw new IllegalStateException("[txn-verify] РАСХОЖДЕНИЕ: read_uncommitted должен быть строго больше read_committed");
        }
        System.out.printf("[txn-verify] OK: read_committed == закоммиченное (%d), read_uncommitted (%d) строго больше — абортнутый батч физически записан, но невидим read_committed-консьюмеру%n",
                committedCount, uncommittedCount);
    }
}

File: messaging/kafka/go/eos/txn.go
package main

import (
	"context"
	"fmt"
	"log"
	"time"

	"github.com/twmb/franz-go/pkg/kgo"
)

// (1)/(2) ядро транзакций: транзакционный producer (BeginTransaction/
// EndTransaction), батч A коммитится штатно, батч B на первой попытке
// абортится (симулированный сбой обработки в середине батча), затем
// пересылается повторно и коммитится. Зеркало
// ../../../java-deep-dive/messaging/.../EosDemo.java — тот же сценарий,
// franz-go API вместо KafkaProducer.
//
// ⚠️ В franz-go НЕТ отдельного вызова initTransactions() — производитель с
// TransactionalID(...) восстанавливает/инициализирует producer ID неявно
// внутри первого BeginTransaction() (maybeRecoverProducerID -> InitProducerId
// под капотом). В Java этот шаг явный (producer.initTransactions()) и
// вызывается один раз до первой транзакции. Функционально эквивалентно —
// разница чисто в форме API, framework-level протокол (InitProducerIdRequest)
// один и тот же.
func newTxnProducer(seeds []string, txnID string) *kgo.Client {
	cl, err := kgo.NewClient(
		kgo.SeedBrokers(seeds...),
		kgo.TransactionalID(txnID),
		kgo.TransactionTimeout(30*time.Second),
		kgo.RequiredAcks(kgo.AllISRAcks()),
		// ⚠️ Идемпотентность НЕЛЬЗЯ выключить при заданном TransactionalID —
		// franz-go возвращает ошибку конфигурации при попытке совместить
		// TransactionalID(...) с DisableIdempotentWrite() ("cannot both
		// disable idempotent writes and use transactional IDs", см.
		// pkg/kgo/config.go). Идемпотентный producer — это подмножество
		// транзакционного, а не отдельная настройка, которую можно забыть
		// включить (в отличие от Java, где enable.idempotence нужно явно
		// выставить true — хотя с 3.0+ там тоже дефолт true).
	)
	if err != nil {
		log.Fatalf("kgo.NewClient (txn producer): %v", err)
	}
	return cl
}

type txnBatchResult struct {
	sentPhysically   int // сколько записей реально ушло в лог партиций (включая абортнутые)
	committedLogical int // сколько сообщений логически подтверждено
}

// sendBatch — один батч в одной транзакции: begin, n записей (опционально
// падаем в середине), commit либо abort. failMidway=true эмулирует сбой
// обработки на 3-й записи батча (индекс 2 из batchSize=5, как в
// java-deep-dive EosDemo) — записи 0,1,2 уже отправлены в кластер (физически
// попадут в лог партиций), но помечаются control-record'ом ABORT.
func sendBatch(cl *kgo.Client, topic, prefix string, batchSize int, failMidway bool) (sent int, committed bool) {
	if err := cl.BeginTransaction(); err != nil {
		log.Fatalf("BeginTransaction (%s): %v", prefix, err)
	}
	ctx := context.Background()
	var produced int
	var midwayErr error
	for i := 0; i < batchSize; i++ {
		rec := &kgo.Record{Topic: topic, Key: []byte(fmt.Sprintf("%s-key-%d", prefix, i)), Value: []byte(fmt.Sprintf("%s-%d", prefix, i))}
		res := cl.ProduceSync(ctx, rec)
		if err := res.FirstErr(); err != nil {
			log.Fatalf("ProduceSync (%s, i=%d): %v", prefix, i, err)
		}
		produced++
		if failMidway && i == 2 {
			midwayErr = fmt.Errorf("симулированный сбой обработки в середине батча %s", prefix)
			break
		}
	}
	if midwayErr != nil {
		if err := cl.AbortBufferedRecords(ctx); err != nil {
			log.Fatalf("AbortBufferedRecords (%s): %v", prefix, err)
		}
		if err := cl.EndTransaction(ctx, kgo.TryAbort); err != nil {
			log.Fatalf("EndTransaction(abort) (%s): %v", prefix, err)
		}
		fmt.Printf("  [txn] батч %s: сбой (%s) -> EndTransaction(TryAbort) — %d записей физически ушли, логически ничего не подтверждено\n", prefix, midwayErr, produced)
		return produced, false
	}
	// EndTransaction(TryCommit) само по себе не флашит — по документации
	// библиотеки Flush должен быть вызван до него явно (ProduceSync уже
	// гарантирует, что каждая отдельная запись подтверждена брокером, но
	// явный Flush — рекомендованная библиотекой практика перед commit).
	if err := cl.Flush(ctx); err != nil {
		log.Fatalf("Flush (%s): %v", prefix, err)
	}
	if err := cl.EndTransaction(ctx, kgo.TryCommit); err != nil {
		log.Fatalf("EndTransaction(commit) (%s): %v", prefix, err)
	}
	fmt.Printf("  [txn] батч %s: EndTransaction(TryCommit) — %d сообщений подтверждено\n", prefix, produced)
	return produced, true
}

func runTxnBatches(seeds []string, topic string, batchSize int) txnBatchResult {
	cl := newTxnProducer(seeds, "cookbook-eos-txn-producer")
	defer cl.Close()

	var result txnBatchResult

	// Батч A: штатный коммит.
	n, ok := sendBatch(cl, topic, "batchA", batchSize, false)
	result.sentPhysically += n
	if ok {
		result.committedLogical += n
	}

	// Батч B, попытка 1: падает в середине -> abort.
	n, ok = sendBatch(cl, topic, "batchB", batchSize, true)
	result.sentPhysically += n
	if ok {
		result.committedLogical += n
	}

	// Батч B, попытка 2 (повтор после "устранения сбоя"): штатный коммит.
	n, ok = sendBatch(cl, topic, "batchB", batchSize, false)
	result.sentPhysically += n
	if ok {
		result.committedLogical += n
	}

	fmt.Printf("[txn] физически отправлено записей: %d, логически подтверждено: %d\n",
		result.sentPhysically, result.committedLogical)
	return result
}

// consumeIsolation читает topic с начала указанным isolation level, пока не
// пройдёт idleTimeout без новых записей. Используется дважды в txn-verify:
// read_committed (должен видеть ровно committedLogical) и read_uncommitted
// (должен видеть sentPhysically, включая абортнутый батч).
func consumeIsolation(seeds []string, topic string, level kgo.IsolationLevel, idleTimeout time.Duration) int {
	cl, err := kgo.NewClient(
		kgo.SeedBrokers(seeds...),
		kgo.ConsumeTopics(topic),
		kgo.ConsumeResetOffset(kgo.NewOffset().AtStart()),
		kgo.FetchIsolationLevel(level),
	)
	if err != nil {
		log.Fatalf("kgo.NewClient (consumer isolation): %v", err)
	}
	defer cl.Close()

	overall, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	count := 0
	lastProgress := time.Now()
	for {
		if overall.Err() != nil {
			break
		}
		pollCtx, pollCancel := context.WithTimeout(overall, 2*time.Second)
		fetches := cl.PollFetches(pollCtx)
		pollCancel()
		n := 0
		fetches.EachRecord(func(r *kgo.Record) { n++ })
		count += n
		if n > 0 {
			lastProgress = time.Now()
		} else if time.Since(lastProgress) > idleTimeout {
			break
		}
	}
	return count
}

func runTxnVerify(seeds []string, topic string, expectCommitted, expectPhysical int) {
	committedCount := consumeIsolation(seeds, topic, kgo.ReadCommitted(), 5*time.Second)
	uncommittedCount := consumeIsolation(seeds, topic, kgo.ReadUncommitted(), 5*time.Second)
	fmt.Printf("[txn-verify] read_committed=%d read_uncommitted=%d (ожидалось: committed=%d physical=%d)\n",
		committedCount, uncommittedCount, expectCommitted, expectPhysical)

	if committedCount != expectCommitted {
		log.Fatalf("[txn-verify] РАСХОЖДЕНИЕ: read_committed=%d, ожидалось %d", committedCount, expectCommitted)
	}
	if uncommittedCount != expectPhysical {
		log.Fatalf("[txn-verify] РАСХОЖДЕНИЕ: read_uncommitted=%d, ожидалось %d", uncommittedCount, expectPhysical)
	}
	if !(uncommittedCount > committedCount) {
		log.Fatalf("[txn-verify] РАСХОЖДЕНИЕ: read_uncommitted (%d) должен быть строго больше read_committed (%d) — абортнутый батч физически на диске, но невидим read_committed", uncommittedCount, committedCount)
	}
	fmt.Printf("[txn-verify] OK: read_committed == закоммиченное (%d), read_uncommitted (%d) строго больше — абортнутый батч физически записан, но невидим read_committed-консьюмеру\n", committedCount, uncommittedCount)
}
